### PR TITLE
Snapshots of persistent VM instances

### PR DESCRIPTION
**Pull request by claudiolor**
# Description

This PR is aimed to introduce the possibility to perform snapshots of persistent VMs. This might allow to install some custom software on the virtual machine, and from that creating a new image to be used in order to create other virtual machines. 
The idea is adding an additional control loop in the instance-operator which perform reconciliation of a new CR called InstanceSnapshot. Once this resource si created, the controller launch a new job in charge of stealing the DataVolume to the persistent VM and creating a new docker image to be pushed to the registry. 

# How Has This Been Tested?

Test suite is provided.

### CODE REVIEW

File: operators/api/v1alpha2/instancesnapshot_types.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// SnapshotStatus is an enumeration representing the current state of the InstanceSnapshot.
+type SnapshotStatus string
+
+const (
+	// Pending -> The request have been created but the snapshot
+	// is waiting to be created.
```
Comment (giorio94) [inline operators/api/v1alpha2/instancesnapshot_types.go:31]:
```suggestion
	// Pending -> The request has been created but the snapshot
	// is waiting to be created.
```

File: operators/api/v1alpha2/instancesnapshot_types.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+	Pending SnapshotStatus = "pending"
```
Comment (giorio94) [inline operators/api/v1alpha2/instancesnapshot_types.go:32]: I would change the first letter of each string to be uppercase: pending -> Pending

File: operators/api/v1alpha2/instancesnapshot_types.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+	// Processing -> The process of creation of the snapshot started.
+	Processing SnapshotStatus = "processing"
+	// Completed -> The snapshot of the instance has been created.
+	Completed SnapshotStatus = "completed"
+	// Failed -> Unfortunately the process of creation of the snapshot failed.
```
Comment (giorio94) [inline operators/api/v1alpha2/instancesnapshot_types.go:37]:
```suggestion
	// Failed -> The process of creation of the snapshot failed.
```

File: operators/api/v1alpha2/instancesnapshot_types.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+// InstanceSnapshotSpec defines the desired state of InstanceSnapshot.
+type InstanceSnapshotSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// +kubebuilder:validation:Required
+
+	// Instance is the reference to the persistent VM instance to be snapshotted
+	// the instance should not be running, otherwise it won't be possible to
+	// steal the volume and extract its content.
```
Comment (giorio94) [inline operators/api/v1alpha2/instancesnapshot_types.go:50]:
```suggestion
	// Instance is the reference to the persistent VM instance to be snapshotted.
	// The instance should not be running, otherwise it won't be possible to
	// steal the volume and extract its content.
```

File: operators/api/v1alpha2/instancesnapshot_types.go [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+	Instance GenericRef `json:"instance.crownlabs.polito.it/InstanceRef"`
+
+	// A template contains a list of environments, this generalize the concept of template and allow to spawn
+	// different vm or containers from the same template.
+	// However, at the moment this functionality has not been implemented and for each template there is one single environment.
+	// The Environment field represent the reference to the environment to be snapshotted, in order to make it compatible
+	// with future upgrades. If not specified, the first available environment is taken.
```
Comment (giorio94) [inline operators/api/v1alpha2/instancesnapshot_types.go:57]: I would just maintain this part of the description, as the previous lines are not strictly related to this resource.

```suggestion
	// Environment represents the reference to the environment to be snapshotted, in case more are
	// associated with the same Instance. If not specified, the first available environment is considered.
```

File: operators/api/v1alpha2/instancesnapshot_types.go [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+	Environment GenericRef `json:"environment.crownlabs.polito.it/EnvironmentRef"`
+
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength.=1
```
Comment (giorio94) [inline operators/api/v1alpha2/instancesnapshot_types.go:61]:
```suggestion
	// +kubebuilder:validation:MinLength=1
```

I suppose this is the correct definition.

I was wondering whether it may be better to force the string in the format "folder/name" as typical for docker images. Otherwise, it may be possible to specify only the name and automatically select the folder (course name? creator?) @palexster WDYT?

File: operators/api/v1alpha2/instancesnapshot_types.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+	// ImageName is the name of the image to pushed in the docker registry.
+	ImageName string `json:"image-name"`
+}
+
+// InstanceSnapshotStatus defines the observed state of InstanceSnapshot.
+type InstanceSnapshotStatus struct {
+	// Phase represent current state of the creation of the vm instance snapshot
```
Comment (giorio94) [inline operators/api/v1alpha2/instancesnapshot_types.go:69]:
```suggestion
	// Phase represents the current state of the Instance Snapshot.
```

File: operators/api/v1alpha2/instancesnapshot_types.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+	// Failed -> Unfortunately the process of creation of the snapshot failed.
+	Failed SnapshotStatus = "failed"
+)
+
+// InstanceSnapshotSpec defines the desired state of InstanceSnapshot.
+type InstanceSnapshotSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// +kubebuilder:validation:Required
```
Comment (giorio94) [inline operators/api/v1alpha2/instancesnapshot_types.go:46]: IIRC, this can be omitted, as it is the default.

File: operators/api/v1alpha2/instancesnapshot_types.go [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+	// the instance should not be running, otherwise it won't be possible to
+	// steal the volume and extract its content.
+	Instance GenericRef `json:"instance.crownlabs.polito.it/InstanceRef"`
+
+	// A template contains a list of environments, this generalize the concept of template and allow to spawn
+	// different vm or containers from the same template.
+	// However, at the moment this functionality has not been implemented and for each template there is one single environment.
+	// The Environment field represent the reference to the environment to be snapshotted, in order to make it compatible
+	// with future upgrades. If not specified, the first available environment is taken.
+	Environment GenericRef `json:"environment.crownlabs.polito.it/EnvironmentRef"`
```
Comment (giorio94) [inline operators/api/v1alpha2/instancesnapshot_types.go:58]:
```suggestion
	Environment GenericRef `json:"environment.crownlabs.polito.it/EnvironmentRef",omitempty`
```

Otherwise it is marked as mandatory in the CRD

File: operators/deploy/instance-operator/templates/clusterrole.yaml [hunk 10]
```diff
@@ -25,6 +25,10 @@ rules:
   resources: ["deployments"]
   verbs: ["get","list","watch","create","patch","update"]
 
+- apiGroups: ["batch", "extensions"]
+  resources: ["jobs", "jobs/status"]
+  verbs: ["get", "list", "watch", "create", "update", "patch", "delete"]
```
Comment (giorio94) [inline operators/deploy/instance-operator/templates/clusterrole.yaml:30]: I would just keep the batch API since it is newer, and drop the delete verb since it is not needed given the deletion is managed through the ownerreference.

File: operators/deploy/instance-operator/templates/clusterrole.yaml [hunk 11]
```diff
@@ -6,7 +6,7 @@ metadata:
     {{- include "instance-operator.labels" . | nindent 4 }}
 rules:
 - apiGroups: ["crownlabs.polito.it"]
-  resources: ["instances", "instances/status"]
+  resources: ["instances", "instances/status", "instancesnapshots", "instancesnapshots/status"]
   verbs: ["get","list","watch","create","update","patch","delete", "deleteCollection"]
```
Comment (giorio94) [inline operators/deploy/instance-operator/templates/clusterrole.yaml:10]: I would drop also here the delete and deleteCollection verbs as not needed

**Review by giorio94 — COMMENTED** [review]
Some minor comments about the API/RBACs. Besides that, they look ok to me.

File: operators/pkg/instancesnapshot-controller/instancesnapshot_helpers.go [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+	templateName := types.NamespacedName{
+		Namespace: instance.Spec.Template.Namespace,
+		Name:      instance.Spec.Template.Name,
+	}
+	template := &crownlabsv1alpha2.Template{}
+
+	if err := r.Get(ctx, templateName, template); err != nil && errors.IsNotFound(err) {
+		// The declared template does not exist set the phase as failed and don't try again.
+		return false, fmt.Errorf("instance %s not found in namespace %s. It is not possible to complete the InstanceSnapshot %s",
+			instanceName.Name, instanceName.Namespace, isnap.Name)
```
Comment (giorio94) [inline operators/pkg/instancesnapshot-controller/instancesnapshot_helpers.go:90]: I suppose instance should be changed to template

File: operators/pkg/instancesnapshot-controller/instancesnapshot_helpers.go [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+	}
+
+	// Check if the environment is a persistent VM.
+	if env.EnvironmentType != crownlabsv1alpha2.ClassVM || !env.Persistent {
+		return false, fmt.Errorf("environment %s is not a persistent VM. It is not possible to complete the InstanceSnapshot %s",
+			env.Name, isnap.Name)
+	}
+
+	// Check if the VM is running.
+	if instance.Spec.Running || instance.Status.Phase != "VmiOff" {
```
Comment (giorio94) [inline operators/pkg/instancesnapshot-controller/instancesnapshot_helpers.go:123]: Here I would just check the Running spec, as the management of the Phase status has some known problems and may not be that reliable

File: operators/pkg/instancesnapshot-controller/instancesnapshot_helpers.go [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+
+	// Define volumes.
+
+	// Define VM VolumeSource.
+	persistentvolsrc := corev1.PersistentVolumeClaimVolumeSource{
+		ClaimName: volumename,
+	}
+	vmvolume := corev1.VolumeSource{
+		PersistentVolumeClaim: &persistentvolsrc,
+	}
```
Comment (giorio94) [inline operators/pkg/instancesnapshot-controller/instancesnapshot_helpers.go:158]: If you want those two can be merged together like (the same for the others):
```suggestion
	vmvolume := corev1.VolumeSource{
		PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
		         ClaimName: volumename,
	        },
	}
```

File: operators/pkg/instancesnapshot-controller/instancesnapshot_helpers.go [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+			VolumeSource: secretvol,
+		},
+	}
+
+	// Define containers.
+
+	// Define Docker pusher container.
+	pushcontainer := corev1.Container{
+		Name:  "docker-pusher",
+		Image: "gcr.io/kaniko-project/executor:latest",
```
Comment (giorio94) [inline operators/pkg/instancesnapshot-controller/instancesnapshot_helpers.go:200]: Never use latest in production, as it is a recipe for disaster. The best would be to add a new flag to the operator to configure it

File: operators/pkg/instancesnapshot-controller/instancesnapshot_helpers.go [hunk 16]
```diff
@@ -0,0 +1,253 @@
...
+				MountPath: "/kaniko/.docker/",
+			},
+		},
+	}
+
+	// Define image exporter container.
+	exportcontainer := corev1.Container{
+		Name: "img-generator",
+		// TODO replace image
+		Image: "claudiolor/img-exporter",
```
Comment (giorio94) [inline operators/pkg/instancesnapshot-controller/instancesnapshot_helpers.go:219]: The image can be built adding an entry as [here](https://github.com/netgroup-polito/CrownLabs/blob/d9a964a5da598b2564545afbfaf35cdfaacc1b09/.github/workflows/build.yml#L124-L126)
Then the name/version should be parametrized as done for the container sidecars

File: operators/pkg/instancesnapshot-controller/instancesnapshot_helpers.go [hunk 17]
```diff
@@ -0,0 +1,253 @@
...
+			return true, c.Type
+		}
+	}
+
+	// Job did not complete.
+	return false, ""
+}
+
+// GetSnapshottingJob generates the job to be created.
+func (r *InstanceSnapshotReconciler) GetSnapshottingJob(isnap *crownlabsv1alpha2.InstanceSnapshot, job *batch.Job) {
```
Comment (giorio94) [inline operators/pkg/instancesnapshot-controller/instancesnapshot_helpers.go:144]: I would change the name of this function to something like CreateSnapshottingJobDefinition, to avoid confusing it with one returning the object from the cluster. I would feel more natural the job to be returned instead of being a parameter

File: operators/pkg/instancesnapshot-controller/instancesnapshot_controller.go [hunk 18]
```diff
@@ -0,0 +1,176 @@
...
+	// the state of its assigned job.
+	jobName := types.NamespacedName{
+		Namespace: snapjob.Namespace,
+		Name:      snapjob.Name,
+	}
+	found := &batch.Job{}
+	err := r.Get(ctx, jobName, found)
+
+	switch {
+	case err != nil && errors.IsNotFound(err):
```
Comment (giorio94) [inline operators/pkg/instancesnapshot-controller/instancesnapshot_controller.go:85]: I would move the logic of each case to a separate function, for the sake of readability/maintainability

File: operators/pkg/instancesnapshot-controller/instancesnapshot_controller_suite_test.go [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+	// Generate whitelist map for InstanceSnapshot controller reconciliation
+	whiteListMap := map[string]string{
+		"test-suite": "true",
+	}
+
+	err = (&instancesnapshot_controller.InstanceSnapshotReconciler{
+		Client:             k8sManager.GetClient(),
+		Scheme:             k8sManager.GetScheme(),
+		EventsRecorder:     k8sManager.GetEventRecorderFor("instance-snapshot"),
+		NamespaceWhitelist: metav1.LabelSelector{MatchLabels: whiteListMap, MatchExpressions: []metav1.LabelSelectorRequirement{}},
```
Comment (giorio94) [inline operators/pkg/instancesnapshot-controller/instancesnapshot_controller_suite_test.go:68]: Here you should add the DeferHook logic as [here](https://github.com/netgroup-polito/CrownLabs/blob/d9a964a5da598b2564545afbfaf35cdfaacc1b09/operators/pkg/instance-controller/suite_test.go#L101) to avoid issues in case of panics during tests.

**Review by giorio94 — COMMENTED** [review]
Looks mostly ok to me. Just some comments inline. I still have to review the tests as WIP

File: operators/deploy/instance-operator/templates/deployment.yaml [hunk 20]
```diff
@@ -47,6 +47,10 @@ spec:
             - "--container-env-vnc-img={{ .Values.configurations.containerEnvironmentOptions.vncImage }}"
             - "--container-env-websockify-img={{ .Values.configurations.containerEnvironmentOptions.websockifyImage }}"
             - "--container-env-novnc-img={{ .Values.configurations.containerEnvironmentOptions.novncImage }}"
+            - "--vm-registry={{ .Values.configurations.vmRegistry.url }}"
+            - "--vm-registry-secret={{ .Values.configurations.vmRegistry.secretName }}"
+            - "--container-export-img={{ .Values.configurations.imgExporter.imgName }}"
```
Comment (giorio94) [inline operators/deploy/instance-operator/templates/deployment.yaml:52]: This needs to be updated to leverage also the correct tag as [here](https://github.com/netgroup-polito/CrownLabs/blob/53ee64928941769aad33079baa2c2c8d928b2733/operators/deploy/instance-operator/templates/deployment.yaml#L46)

File: operators/pkg/instancesnapshot-controller/instancesnapshot_logic.go [hunk 21]
```diff
@@ -0,0 +1,75 @@
...
+
+	isnap.Status.Phase = crownlabsv1alpha2.Processing
+	if err1 := r.Status().Update(ctx, isnap); err1 != nil {
+		return true, fmt.Errorf("error when updating status of InstanceSnapshot %s -> %w", isnap.Name, err1)
+	}
+
+	return false, nil
+}
+
+// HandleExistingJob checks the status of the existing job and update the status of the InstanceSnapshot accordingly.
```
Comment (giorio94) [inline operators/pkg/instancesnapshot-controller/instancesnapshot_logic.go:56]:
```suggestion
// HandleExistingJob checks the status of the existing job and updates the status of the InstanceSnapshot accordingly.
```

File: operators/pkg/instancesnapshot-controller/instancesnapshot_controller_test.go [hunk 22]
```diff
@@ -0,0 +1,335 @@
+package instancesnapshot_controller_test
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/rand"
+
```
Comment (giorio94) [inline operators/pkg/instancesnapshot-controller/instancesnapshot_controller_test.go:10]: This empty line should be dropped, as it prevents correct ordering

File: operators/pkg/instancesnapshot-controller/instancesnapshot_controller_test.go [hunk 23]
```diff
@@ -0,0 +1,335 @@
...
+		}
+		return retrievedIsnap.Status.Phase
+	}, timeout, interval).Should(Equal(desiredStatus))
+}
+
+func checkIsnapSuccessfulCreation(ctx context.Context, isnap *crownlabsv1alpha2.InstanceSnapshot, workingNamespace string, timeout, interval time.Duration) {
+	By("Creating the InstanceSnapshot resource")
+	Expect(k8sClient.Create(ctx, isnap)).Should(Succeed())
+
+	By("Checking that the instance has been created")
```
Comment (giorio94) [inline operators/pkg/instancesnapshot-controller/instancesnapshot_controller_test.go:305]:
```suggestion
	By("Checking that the instance snapshot has been created")
```

File: operators/pkg/instancesnapshot-controller/instancesnapshot_controller_test.go [hunk 24]
```diff
@@ -0,0 +1,335 @@
...
+		})
+
+		It("Should fail: template does not exist", func() {
+			By("Getting current instance")
+			currentInstance := &crownlabsv1alpha2.Instance{}
+			instanceLookupKey := types.NamespacedName{Name: InstanceName, Namespace: WorkingNamespace}
+			Expect(k8sClient.Get(ctx, instanceLookupKey, currentInstance)).Should(Succeed())
+
+			By("Changing template with a non-existing one")
+			currentInstance.Spec.Template.Name = "blabla"
```
Comment (giorio94) [inline operators/pkg/instancesnapshot-controller/instancesnapshot_controller_test.go:251]:
```suggestion
			currentInstance.Spec.Template.Name = "invalid"
```

File: deploy/crownlabs/values.yaml [hunk 25]
```diff
@@ -43,6 +43,14 @@ instance-operator:
       novncImage: crownlabs/novnc
       filebrowserImage: filebrowser/filebrowser
       filebrowserImageTag: latest
+    kaniko:
+      tag: "v1.5.2"
+      imageUrl: "gcr.io/kaniko-project/executor"
+    vmRegistry:
+      url: registry.internal.crownlabs.polito.it
+      secretName: registry-credentials
+    imgExporter:
```
Comment (giorio94) [inline deploy/crownlabs/values.yaml:52]: Would group this and the kanico entry together, as strictly related

**Review by giorio94 — COMMENTED** [review]
Had a look through your PR and looks mostly ready to me.

Just a couple of nits inline

File: operators/pkg/instancesnapshot-controller/instancesnapshot_helpers.go [hunk 26]
```diff
@@ -0,0 +1,237 @@
...
+		Name:      isnap.Spec.Instance.Name,
+	}
+	instance := &crownlabsv1alpha2.Instance{}
+
+	if err := r.Get(ctx, instanceName, instance); err != nil {
+		return batch.Job{}, fmt.Errorf("error in retrieving the instance for InstanceSnapshot %s -> %w", isnap.Name, err)
+	}
+
+	var backoff int32 = 2
+	imagetag := fmt.Sprint(time.Now().Format("02012006t150405"))
```
Comment (giorio94) [inline operators/pkg/instancesnapshot-controller/instancesnapshot_helpers.go:132]: Could you reverse the order of the date to ensure ordering? (i.e. YYYYMMDD)

File: operators/pkg/instancesnapshot-controller/instancesnapshot_controller.go [hunk 27]
```diff
@@ -0,0 +1,153 @@
...
+		defer r.ReconcileDeferHook()
+	}
+
+	isnap := &crownlabsv1alpha2.InstanceSnapshot{}
+
+	if err := r.Get(ctx, req.NamespacedName, isnap); client.IgnoreNotFound(err) != nil {
+		klog.Errorf("Error when getting InstanceSnapshot %s before starting reconcile -> %s", isnap.Name, err)
+		return ctrl.Result{}, err
+	} else if err != nil {
+		klog.Infof("InstanceSnapshot %s already deleted", isnap.Name)
```
Comment (giorio94) [inline operators/pkg/instancesnapshot-controller/instancesnapshot_controller.go:75]: Since an error occurred, isnap.Name is empty. You should probably use req.NamespacedName.name

**Review by giorio94 — COMMENTED** [review]
Some nits inline

**Review by giorio94 — APPROVED** [review]
/lgtm

Would just discuss tomorrow about kanico, before merging

File: .github/workflows/build.yml [hunk 28]
```diff
@@ -156,6 +156,10 @@ jobs:
           - component: ssh-bastion
             context: ./operators/build/ssh-bastion
 
+          # img-exporter image for InstanceSnapshot
```
Comment (GabriFila) [inline .github/workflows/build.yml:159]: @giorio94 shouldn't there be also a line with `img-exporter` before, e.g. at line 78 of this file?

Comment (giorio94) [reply .github/workflows/build.yml:159]: Definitely. Thanks, I missed that.

**Review by GabriFila — APPROVED** [review]
